src/main/java/org/gaul/s3proxy/AwsSignature.java

```diff
@@ -114,9 +116,23 @@ static String createAuthorizationSignature(
                         HttpHeaders.CONTENT_TYPE)))
                 .append('\n');
         String expires = request.getParameter("Expires");
-        if (expires != null) {
-            builder.append(expires);
-        } else if (!canonicalizedHeaders.containsKey("x-amz-date")) {
+        if (queryauth) {
+          // If expires is  not nil, then it is query string sign
+          // If expires is nil,maybe alse query string sign
+          // So should check other accessid para ,presign to judge.
+          // not the expires
+            builder.append(Strings.nullToEmpty(expires));
+
+        } else if (!bothdateheader &&
+            canonicalizedHeaders.containsKey("x-amz-date")) {
+          // If x-amz-date is nil, use date .
+          // If date and x-amz-date both not nil, date=x-amz-date,no x-amz-date
+          // If amz date not nil,date nil,this case
+            builder.append("");
+        } else if (bothdateheader &&
+            !canonicalizedHeaders.containsKey("x-amz-date")) {
```

Comment:
I don't understand this logic -- if you have both date headers then you will also have `x-amz-date`.  Instead should you have a single check for `if (bothDateHeader)` followed by individual checks for each header?

Reply:
```
         else if (!bothdateheader &&                                                                                                                                                                        
            canonicalizedHeaders.containsKey("x-amz-date")) {
             // If amz date not nil,date nil,this case
              builder.append("");
          } else if (bothdateheader &&
              !canonicalizedHeaders.containsKey("x-amz-date")) {
               // If date and x-amz-date both not nil, date=x-amz-date,no x-amz-date 
              builder.append(request.getHeader("x-amz-date"));
          } else {
             //  If x-amz-date is nil, use date .

              builder.append(request.getHeader(HttpHeaders.DATE));
           }
```

Reply:
I will think about it

Reply:
When have both dataheaders, we will aslo have x-amz-date header.But we once create cannicalizeheaders,we ignor x-amz-date header in createAuthorizationSignature(v2).

Reply:
When only have x-amz-header, need not to add date ,only process the canonicalizedHeaders which have the header.
If x-amz-date is nil, use date header.